src/range/byte_content_range.rs

```diff
@@ -0,0 +1,327 @@
+/// #
+/// use http_types::range::{ByteContentRange, ByteRange};
+/// use http_types::{Response, StatusCode};
+///
+/// let mut content_range = ByteContentRange::new().with_range(1, 5);
+///
+/// let mut res = Response::new(StatusCode::PartialContent);
+/// content_range.apply(&mut res);
+///
+/// let content_range = ByteContentRange::from_headers(res)?.unwrap();
+/// assert_eq!(content_range.range(), Some(ByteRange::new(1, 5)));
+/// assert_eq!(content_range.size(), None);
+/// #
+/// # Ok(()) }
+/// ```
+///
+/// Responding to an invalid range request for a 10 bytes document:
+///
+/// ```
+/// # fn main() -> http_types::Result<()> {
+/// #
+/// use http_types::range::{ByteContentRange, ByteRange};
+/// use http_types::{Response, StatusCode};
+///
+/// let mut content_range = ByteContentRange::new().with_size(10);
+///
+/// let mut res = Response::new(StatusCode::RequestedRangeNotSatisfiable);
+/// content_range.apply(&mut res);
+///
+/// let content_range = ByteContentRange::from_headers(res)?.unwrap();
+/// assert_eq!(content_range.range(), None);
+/// assert_eq!(content_range.size(), Some(10));
+/// #
+/// # Ok(()) }
+/// ```
+#[derive(Default, Debug, Clone, Eq, PartialEq)]
+pub struct ByteContentRange {
+    size: Option<u64>,
+    range: Option<ByteRange>,
+}
+
+impl ByteContentRange {
+    const CONTENT_RANGE_PREFIX: &'static str = "bytes";
+
+    /// Create a new instance with no range and no size.
+    pub fn new() -> Self {
+        ByteContentRange::default()
+    }
+
+    /// Returns a new instance with a given range defined by `start` and `end` bounds.
+    pub fn with_range(mut self, start: u64, end: u64) -> Self {
+        self.range = Some(ByteRange::new(start, end));
+        self
+    }
+
+    /// Returns a new instance with a size.
+    pub fn with_size(mut self, size: u64) -> Self {
+        self.size = Some(size);
+        self
+    }
+
+    /// Returns the `ByteRange` if any.
+    pub fn range(&self) -> Option<ByteRange> {
+        self.range
+    }
+
+    /// Returns the size if any.
+    pub fn size(&self) -> Option<u64> {
+        self.size
+    }
+
+    /// Create a new instance from a Content-Range headers.
+    ///
+    /// Only a single Content-Range per resource is assumed to exist. If multiple Range
+    /// headers are found the last one is used.
+    pub fn from_headers(headers: impl AsRef<Headers>) -> crate::Result<Option<Self>> {
+        let headers = match headers.as_ref().get(CONTENT_RANGE) {
+            Some(headers) => headers,
+            None => return Ok(None),
+        };
+
+        // If a header is returned we can assume at least one exists.
+        let s = headers.iter().last().unwrap().as_str();
+        if !s.starts_with(Self::CONTENT_RANGE_PREFIX) {
+            return Ok(None);
+        }
+        Self::from_str(s).map(Some)
+    }
+
+    /// Create a ByteRanges from a string.
+    pub(crate) fn from_str(s: &str) -> crate::Result<Self> {
+        let fn_err = || {
```

Comment:
There is no error code specified for invalid `Content-Range` header in a response. I chose to use the error the same error code as the one used for bad range requests.